[WFLY-9781] created simple integration test for Forwarded handler.

jstourac opened the pull request:
Created integration tests for https://issues.jboss.org/browse/WFLY-9781.

jstourac committed:
[WFLY-9781] created simple integration test for Forwarded handler.

stuartwdouglas commented on testsuite/integration/web/src/test/java/org/jboss/as/test/integration/web/handlers/ForwardedTestHelperHandler.java:
```diff
@@ -0,0 +1,24 @@
+
+
+    public void handleRequest(HttpServerExchange exchange) throws Exception {
+        exchange.getResponseSender().send(exchange.getSourceAddress() + "|" + exchange.getRequestScheme() + "|"
```
This is not really valid, after you have used the response sender in this way the exchange will be ended, so the next handler will be operating on a completed exchange, if you want to send back this information you should do it in a header.

stuartwdouglas commented on testsuite/integration/web/src/test/java/org/jboss/as/test/integration/web/handlers/ForwardedHandlerTestCase.java:
```diff
@@ -0,0 +1,124 @@
+    private static final String JBOSS_WEB_TEXT = "<?xml version=\"1.0\"?>\n" +
+            "<jboss-web>\n" +
+            "    <http-handler>\n" +
+            "        <class-name>org.jboss.as.test.integration.web.handlers.ForwardedTestHelperHandler</class-name>\n" +
```
This should really be testing with a Servlet, rather than an Undertow handler as that is what 99%+ of people will actually be using.

jstourac replied:
Thanks - I'll extended this so also servlet is utilized.

kabir commented:
@jstourac Adding the fixme label. Please ping me when done

jstourac commented:
@stuartwdouglas I have updated this to address your comments. Although, now two tests that I have added are failing - those are `testRewriteGlobalSettingsServlet` and `testRewriteWithUndertowHandlersServlet`. Both utilize servlet instead of handler. And problem is with  methods `HttpServletRequest.getLocalAddr()` and `HttpServletRequest.getLocalPort()`. These methods actually don't contain what is provided by `Forwarded ~ by` header. But have original local address and port (127.0.0.1 and 8080 in my case).

stuartwdouglas commented:
You should get checking getRemoteAddr not getLocalAddr

jstourac commented:
Actually the test checks also `getRemoteAddr` values. The problem is that `getLocalAddr` and `getLocalName` returns different results.

Header value in request:
```
Forwarded=for=192.121.210.60:455;proto=https;by=203.0.113.43:777
```
What particular methods return in case of handle usage:
```
HttpServerExchange.getSourceAddress() = /192.121.210.60:455
HttpServerExchange.getRequestScheme() = https
HttpServerExchange.getDestinationAddress() = /203.0.113.43:777
```
Those are correct.

Servlet case:
```
HttpServletRequest.getRemoteAddr() = 192.121.210.60
HttpServletRequest.getRemoteHost() = 192.121.210.60
HttpServletRequest.getRemotePort() = 455
HttpServletRequest.getScheme() = https
HttpServletRequest.getLocalName() = 203.0.113.43
HttpServletRequest.getLocalAddr() = 127.0.0.1
HttpServletRequest.getLocalPort() = 8080
```
Values of `getLocalName()` and `getLocalAddr()` should be the same, IMHO.

jstourac commented:
https://issues.jboss.org/browse/UNDERTOW-1280 has fixed the issue with getLocalAddr() returned value. If https://issues.jboss.org/browse/UNDERTOW-1282 is accepted and merged, these tests start to pass.

@stuartwdouglas do you see something I should improve or rewrite in this MR anymore?

jstourac commented:
retest this please

stuartwdouglas commented:
Adding hold label as this needs an updated Undertow version

stuartwdouglas commented:
Retest this please

stuartwdouglas commented:
Retest this please

jstourac commented:
Well, tests failed with:
>*  Cannot find commit 60c30982dc764cc2f51e0f8e6c8c5940c4b18c84 in the https://github.com/wildfly/wildfly.git repository, possible reason: refs/pull/10860/merge branch was updated and the commit selected for the build is not reachable anymore
>* Failed to perform checkout on agent: Cannot find commit 60c30982dc764cc2f51e0f8e6c8c5940c4b18c84 in the https://github.com/wildfly/wildfly.git repository, possible reason: refs/pull/10860/merge branch was updated and the commit selected for the build is not reachable anymore

Not sure I understand the root cause as I have not updated this for a while... Anyway, I tried to rebase. Let's see the results now...

jstourac commented:
Okay, the tests seem to pass after the rebase :)
